From 42224ed7ddf593716663c1ceb01416e89afa76f8 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:04:12 -0400
Subject: [powerpc] Access local paca after hard irq disabled

Message-id: <1410545655-205645-24-git-send-email-dzickus@redhat.com>
Patchwork-id: 94161
O-Subject: [RHEL7 PATCH 023/626] powerpc: Access local paca after hard irq disabled
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 0b88f772bdf2847461debf417b1ee96043a7cb1b
Author: Tiejun Chen <tiejun.chen@windriver.com>
Date:   Mon Jul 15 10:36:04 2013 +0800

    powerpc: Access local paca after hard irq disabled

    In hard_irq_disable(), we accessed the PACA before we hard disabled
    the interrupts, potentially causing a warning as get_paca() will
    us debug_smp_processor_id().

    Move that to after the disabling, and also use local_paca directly
    rather than get_paca() to avoid several redundant and useless checks.

    Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/hw_irq.h b/arch/powerpc/include/asm/hw_irq.h
index ba713f1..10be1dd 100644
--- a/arch/powerpc/include/asm/hw_irq.h
+++ b/arch/powerpc/include/asm/hw_irq.h
@@ -96,10 +96,11 @@ static inline bool arch_irqs_disabled(void)
 #endif
 
 #define hard_irq_disable() do {   \
- u8 _was_enabled = get_paca()->soft_enabled; \
+ u8 _was_enabled;    \
  __hard_irq_disable();    \
- get_paca()->soft_enabled = 0;   \
- get_paca()->irq_happened |= PACA_IRQ_HARD_DIS; \
+ _was_enabled = local_paca->soft_enabled; \
+ local_paca->soft_enabled = 0;   \
+ local_paca->irq_happened |= PACA_IRQ_HARD_DIS; \
  if (_was_enabled)    \
   trace_hardirqs_off();   \
 } while(0)
-- 
1.7.1